src/Microsoft.ML.Transforms/Dracula/Featurizer.cs

```diff
@@ -0,0 +1,224 @@
+            // int: label bin count
+            // int: #features
+            // Single: prior coefficient
+            // Single: laplace scale
+            // _countTable
+
+            ctx.Writer.Write(_labelBinCount);
+            ctx.Writer.Write(NumFeatures);
+            ctx.Writer.WriteSinglesNoCount(PriorCoef);
+            ctx.Writer.WriteSinglesNoCount(LaplaceScale);
+            ctx.SaveModel(_countTables, "CountTables");
+        }
+
+        public int NumFeatures { get; }
+
+        public void GetFeatureNames(string[] classNames, ref VBuffer<ReadOnlyMemory<char>> featureNames)
+        {
+            if (classNames == null)
+            {
+                classNames = new string[_labelBinCount];
+                for (int i = 0; i < _labelBinCount; i++)
+                    classNames[i] = string.Format("Class{0:000}", i);
+            }
+
+            Contracts.Check(Utils.Size(classNames) == _labelBinCount, "incorrect class names");
+
+            var editor = VBufferEditor.Create(ref featureNames, NumFeatures);
+            for (int i = 0; i < _labelBinCount; i++)
+                editor.Values[i] = $"{classNames[i]}_Count".AsMemory();
+            for (int i = 0; i < _logOddsCount; i++)
+                editor.Values[_labelBinCount + i] = $"{classNames[i]}_LogOdds".AsMemory();
+            editor.Values[NumFeatures - 1] = "IsBackoff".AsMemory();
+            featureNames = editor.Commit();
+        }
+
+        public void GetFeatures(int iCol, int iSlot, Random rand, long key, Span<float> features)
+        {
+            _host.Assert(features.Length == NumFeatures);
+
+            // get counts from count table in the first _labelBinCount indices.
+            var countsBuffer = features.Slice(0, _labelBinCount);
+            var countTable = _countTables[iCol, iSlot];
+            countTable.GetCounts(key, countsBuffer);
+
+            // check if it's garbage and replace with garbage counts if true
+            float sum = 0;
+            foreach (var feat in countsBuffer)
+                sum += feat;
+            bool isGarbage = sum < countTable.GarbageThreshold;
+            if (isGarbage)
+            {
+                int i = 0;
+                foreach (var count in countTable.GarbageCounts)
+                    countsBuffer[i++] = count;
+            }
+
+            sum = AddLaplacianNoisePerLabel(iCol, rand, countsBuffer);
+
+            // add log odds in the next _logOddsCount indices.
+            GenerateLogOdds(iCol, countTable, countsBuffer, features.Slice(_labelBinCount, _logOddsCount), sum);
+            AssertValidOutput(features);
+
+            // Add the last feature: an indicator for isGarbage.
+            features[NumFeatures - 1] = isGarbage ? 1 : 0;
+        }
+
+        // Assumes features are filled with the respective counts.
+        // Adds laplacian noise if set and returns the sum of the counts.
+        private float AddLaplacianNoisePerLabel(int iCol, Random rand, Span<float> counts)
```

Comment:
Is the noise still added once the ML model is fully created?

@davidbrownellWork was mentioning that there is added noise at the prediction time. This complicates the ONNX conversion. The AutoML team is looking to use Dracula for its target encoding and we need it in ML.NET, NimbusML, and all the way ONNX.

Reply:
The noise is only used at prediction time. From your experience using this transform, do you think it is needed?

---
In reply to: [356322013](https://github.com/dotnet/machinelearning/pull/4514#discussion_r356322013) [](ancestors = 356322013)

Reply:
I looked at the EstimatorChain class, and it seems to me that it would be possible to add a method similar to AppendCacheCheckPointthat would add noise to a requested column. Here is the code that is currently in the Fit method of EstimatorChain:
```
            IDataView current = input;
            var xfs = new ITransformer[_estimators.Length];
            for (int i = 0; i < _estimators.Length; i++)
            {
                var est = _estimators[i];
                xfs[i] = est.Fit(current);
                current = xfs[i].Transform(current);
                if (_needCacheAfter[i] && i < _estimators.Length - 1)
                {
                    Contracts.AssertValue(_host);
                    current = new CacheDataView(_host, current, null);
                }
            }
```

If we add, similarly to how we have `current = new CacheDataView(...)`, something like `current = new NoiseDataView(...)`, then we can let the user writes something like
```
var pipeline = mlContext.Transforms.CountTargetEncoding()
    .AppendNoise("Features").Append(mlContext.Binary.Sdca());
```

then the SDCA component in the pipeline will see the `Features` column with noise added to it when training, however, the resulting `TransformerChain` will not have noise as part of the model.
 
Together with good samples and documentation emphasizing the recommended way to use the `CountTargetEncodingEstimator` (which is either to train the transformer in a separate pipeline than the trainer and on different data, or use the AppendNoise API), I think that may be a better solution to having a noise option as part of the estimator constructor. 

Please let me know what you think of this solution and whether I should go ahead and implement that.


---
In reply to: [357085107](https://github.com/dotnet/machinelearning/pull/4514#discussion_r357085107) [](ancestors = 357085107,356322013)

Reply:
Hi Yael, what is the name of this transform in TLC? I went to our TLC telemetry to look for internal users of the dracula transform but I did not see anyone.

Reply:
In the TLC GUI it's called "Dracula Transform" (under "Featurizers: Supervised") and in the TLC code it is called `DraculaTransform`… but I wouldn't know if other names are used in the telemetry.

Reply:
If I recall, the log-odds are returned _approximately_ as:
```c#
CountPositve = CountPositive + LaplacianNoise(); // Add noise to pos count
CountNegative = CountNegative + LaplacianNoise();  // Add noise to neg count
logOdds = ln( (CountPositive + PriorCount) / (CountPositive + CountNegative + 2*PriorCount) );  // Includes 2x noise, plus prior as smoothing
```

I don't think we'll be able to just add noise outside of the Dracula transform and end up with the same result.

My recommendation is keeping the noise computation as is, and disable noise for the ONNX unit tests, or add a window  (e.g. +/- 10%)  of acceptable values when checking for (near) equality of the outputs.

Reply:
The two possible solutions as I see it are:
* Leaving the computation as it is right now.
  * Pros: easy solution for fitting a pipeline that has a trainer after the Dracula transform without having to split into two pipelines and training each part on a different dataset.
  * Cons: if the ONNX export of this transform will not include the noise part (which it likely won't), then this will result in a model that gives different results in the ML.NET format and in the ONNX format.
* Making the transformer apply the noise, but not serialize the noise parameter, i.e. after deserialization it won't add noise.
  * Pros: consistency with the ONNX version of the model. Also, the noise is really not needed at prediction time, it is only needed to train the next component in the pipeline.
  * Cons: if you call `Transform` on the pipeline with the same data right after it is trained, and after serializing and deserializing the model, it will give different results.

I don't particularly like either of these solutions, I think they are both problematic, but currently I can't think of a solution that will not have any of the cons listed above...  Which solution do you think I should go with?

cc @justinormont, @eerhardt


---
In reply to: [365470116](https://github.com/dotnet/machinelearning/pull/4514#discussion_r365470116) [](ancestors = 365470116)

Reply:
The introduction of noise is a transformation step that is only used when the featurizer is included in a pipeline that is used during training time. Because ONNX exports are only used for inference, it seems OK that the export doesn't include noise, as the usage in ONNX will be the same as when this featurizer is used in an inferencing ML.Net pipeline.

It seems that the first point is the solution, as long as the code is written such that noise is not introduced when the featurizer is part of an inferencing pipeline.